lib/std/mem.zig

```diff
@@ -196,26 +196,70 @@ test "Allocator.resize" {
 /// `copyForwards` if they do.
 pub const copy = copyForwards;
 
+/// Returns true if the slices overlap
+/// This is an *exclusive* range track
+pub fn slicesOverlap(comptime T: type, slice_a: []const T, slice_b: []const T) bool {
+    // NOTE: We can't currently determine if ranges overlap at comptime (as the
+    // currently language doesn't expose the information needed to do that).
+    const a_begin = @intFromPtr(slice_a.ptr);
+    const b_begin = @intFromPtr(slice_b.ptr);
+
+    // The ptrs to one element beyond the last element of a & B
+    const a_end = @intFromPtr(slice_a.ptr + slice_a.len);
+    const b_end = @intFromPtr(slice_b.ptr + slice_b.len);
+
+    // This coment is written assuming a_begin is > b_begin... the other case is
+    // an exercise for the reader ;)
+    //
+    // The first element of a comes after slice_b, so we need to confirm that
+    // the last element of slice_b doesn't come at or after the first element of
+    // slice_a.
+    // b_end points to 1 element beyond the range of slice_b, so we need to
+    // check that b_end > a_begin - there is no aliasing if b_end == a_begin
+    // because b_end isn't actually part of b_slice itself!
+    if (a_begin > b_begin) {
+        return b_end > a_begin;
+    } else if (b_begin > a_begin) {
+        return a_end > b_begin;
+    }
+    assert(a_begin == b_begin);
+    return true;
```

Comment:
I find all of these comments and logic confusing, when you could do simply instead:

```suggestion
    return !(a_begin >= b_end or b_begin >= a_end);
```

Reply:
🙈  I can overcomplicate simple things sometimes! That makes so much more sense!